### ISSUE TITLE
About WeightedBCEWithLogitLoss

### ISSUE THREAD

**Issue by JialeTao** [RoyalVane/CLAN#11]
Hi, I have  a problem in understanding your code of class WeightedBCEWithLogitLoss, showed as follow. It seems the input is logits, but there isn't any sigmoid or softmax operation done on it. And I'm confused of that two lines. Is the loss computation same as pytorch BCEWithLogitLoss???   
![image](https://user-images.githubusercontent.com/38600167/63083978-b67ce880-bf7c-11e9-85fc-21338e5f0a4e.png)

**Comment by RoyalVane** [comment]
Yes，the designed loss is different from the vanilla BCEWithLogitLoss. Besides the weights,  it is implemented without the sigmoid or softmax operation.

**Comment by JialeTao** [comment]
> Yes，the designed loss is different from the vanilla BCEWithLogitLoss. Besides the weights, it is implemented without the sigmoid or softmax operation.

Hi, I see the codes of weighted mechanism. But does the computaion of that two lines really means crossentropy loss as paper said? I still don't undestand the meanning of that two lines. Could you explain the relationship between the computation and the formular in original paper? If that I'll be appreciated.
![image](https://user-images.githubusercontent.com/38600167/63241601-e7179780-c286-11e9-84e7-eda25654994d.png)

**Comment by RoyalVane** [comment]
A balanced adversarial learning requires the loss of discriminator to converge around 0.5. However, the value of M(P1,P2) tends to be very small in the late epochs, and the discriminator using vanilla BCE loss would tend to bias to the "real samples", resulting in an unbalanced  training course . Such designed loss (without softmax) in this repository aims to guarantee larger gradients to be backpropagated to G and D in even later epochs, thus stablizing the training.  Under the premise of consistent gradient decent direction with paper, such WeightedBCEWithLogitLoss allows a more balanced adversarial course.

**Comment by JialeTao** [comment]
> A balanced adversarial learning requires the loss of discriminator to converge around 0.5. However, the value of M(P1,P2) tends to be very small in the late epochs, and the discriminator using vanilla BCE loss would tend to bias to the "real samples", resulting in an unbalanced training course . Such designed loss (without softmax) in this repository aims to guarantee larger gradients to be backpropagated to G and D in even later epochs, thus stablizing the training. Under the premise of consistent gradient decent direction with paper, such WeightedBCEWithLogitLoss allows a more balanced adversarial course.

Hi, I guess I've undestood the computation of the two lines. It in fact computes the BCEWithLogitLoss  
and contains the sigomoid operation in it. An good  implementation!!! But you said  it is implemented without the sigmoid or softmax operation, I hope I didn't understand wrong. Thanks a lot.
![image](https://user-images.githubusercontent.com/38600167/63267656-206ef800-c2c5-11e9-9598-f423885e7673.png)

**Comment by RoyalVane** [comment]
You are welcome. Thank you again for your interest in CLAN!
